### ISSUE TITLE
`tsc --watch` is watching too many files/directories when `allowJs` is enabled

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#31997]
**TypeScript Version:** 3.5.2

**Search Terms:** watch allowJs triggered tsconfig json

**Code**

When `allowJs` is used, it seems `tsc` watches more files/directories than it needs to. Here is a reduced test case.

``` json
// tsconfig.json
{
  "compilerOptions": {
    "allowJs": true,
  },
  "files": [
    "./a.js"
  ]
}
```

```ts
// ./a.js
import './b';
```

```ts
// ./b.js
export default 1;
```

Now run:

```
tsc --watch --extendedDiagnostics --noEmit
```

After the first compilation is finished, run

```
mkdir ./foo/
```

**Expected behavior:**
Watch should not be triggered because the created folder (`./foo/`) should not be part of the watch.

**Actual behavior:**
Watch is triggered:

```
DirectoryWatcher:: Triggered with /Users/oliverjash/Development/typescript-playground/src/foo :: WatchInfo: /Users/oliverjash/Development/typescript-playground/src 1 Failed Lookup Locations
Elapsed:: 0ms DirectoryWatcher:: Triggered with /Users/oliverjash/Development/typescript-playground/src/foo :: WatchInfo: /Users/oliverjash/Development/typescript-playground/src 1 Failed Lookup Locations
```

**Comment by sheetalkamat** [comment]
Watching `/Users/oliverjash/Development/typescript-playground/src ` isnt unnecessary. Its watching it to see if files like `/Users/oliverjash/Development/typescript-playground/src /b.ts` or `/Users/oliverjash/Development/typescript-playground/src/b.d.ts` are created. Because the watcher is invoked with the directory of no interest no action is taken. Note that we watch directory instead of files to optimize cases for multiple failed lookup locations.

You can run tsc with `--traceResolution` to see what all locations are checked before resolution succeeds.

**Comment by OliverJAsh** [comment]
Thanks @sheetalkamat! For some reason when I disable `allowJs`, I see slightly different behaviour. When `allowJs` disabled, when running with `--extendedDiagnostics` and following the steps above, nothing happens. If I enable `allowJs`, then something does appear to happen (pasted output at bottom of OP).

**Comment by sheetalkamat** [comment]
With non allow js you wouldn't have failed lookup location because the first try of `/Users/oliverjash/Development/typescript-playground/src /b.ts` is going to succeed so no alternatives to watch for. In case of `.js` file `.ts` and `.d.ts` files are preferred so those locations need to watched.
